mcp/src/main/java/io/modelcontextprotocol/spec/McpStreamableServerSession.java

```diff
@@ -0,0 +1,325 @@
+			return notificationHandler.handle(
+					new McpAsyncServerExchange(this.id,
+							listeningStream != null ? listeningStream : MissingMcpTransportSession.INSTANCE,
+							this.clientCapabilities.get(), this.clientInfo.get(), transportContext),
+					notification.params());
+		});
+
+	}
+
+	public Mono<Void> accept(McpSchema.JSONRPCResponse response) {
+		return Mono.defer(() -> {
+			var stream = this.requestIdToStream.get(response.id());
+			if (stream == null) {
+				return Mono.error(new McpError("Unexpected response for unknown id " + response.id())); // TODO
+																										// JSONize
+			}
+			var sink = stream.pendingResponses.remove(response.id());
+			if (sink == null) {
+				return Mono.error(new McpError("Unexpected response for unknown id " + response.id())); // TODO
+																										// JSONize
+			}
+			else {
+				sink.success(response);
+			}
+			return Mono.empty();
+		});
+	}
+
+	record MethodNotFoundError(String method, String message, Object data) {
+	}
+
+	private MethodNotFoundError getMethodNotFoundError(String method) {
+		return new MethodNotFoundError(method, "Method not found: " + method, null);
+	}
+
+	@Override
+	public Mono<Void> closeGracefully() {
+		return Mono.defer(() -> {
+			McpStreamableServerSessionStream listeningStream = this.listeningStreamRef.get();
+			return listeningStream != null ? listeningStream.closeGracefully() : Mono.empty(); // TODO:
+																								// Also
+																								// close
+																								// all
+																								// the
+																								// open
+																								// streams
+		});
+	}
+
+	@Override
+	public void close() {
+		McpStreamableServerSessionStream listeningStream = this.listeningStreamRef.get();
+		if (listeningStream != null) {
+			listeningStream.close();
+		}
+		// TODO: Also close all open streams
+	}
+
+	/**
+	 * Request handler for the initialization request.
+	 */
+	public interface InitRequestHandler {
+
+		/**
+		 * Handles the initialization request.
+		 * @param initializeRequest the initialization request by the client
+		 * @return a Mono that will emit the result of the initialization
+		 */
+		Mono<McpSchema.InitializeResult> handle(McpSchema.InitializeRequest initializeRequest);
+
+	}
+
+	/**
+	 * Notification handler for the initialization notification from the client.
+	 */
+	public interface InitNotificationHandler {
+
+		/**
+		 * Specifies an action to take upon successful initialization.
+		 * @return a Mono that will complete when the initialization is acted upon.
+		 */
+		Mono<Void> handle();
+
+	}
+
+	public interface Factory {
+
+		McpStreamableServerSessionInit startSession(McpSchema.InitializeRequest initializeRequest);
+
+	}
+
+	public record McpStreamableServerSessionInit(McpStreamableServerSession session,
+			Mono<McpSchema.InitializeResult> initResult) {
+	}
+
+	public final class McpStreamableServerSessionStream implements McpSession {
```

Comment:
The `McpStreamableServerSessionStream` isn't bidirectional like `McpSession`. 
We can split this into two interfaces:
- `McpRpcSender` - handles outgoing messages (copies current McpSession functionality)
- `McpRpcReceiver` - handles incoming messages with `handle(JSONRPCMessage)`
(names  be refined)

McpSession would implement both interfaces, McpStreamableServerSessionStream would only implement McpRpcSender, and the Stateless Server would use McpRpcReceiver.